newm-server/src/main/kotlin/io/newm/server/features/marketplace/repo/MarketplaceRepositoryImpl.kt

```diff
@@ -68,6 +100,118 @@ internal class MarketplaceRepositoryImpl(
         }
     }
 
+    override suspend fun generateOrderAmount(request: OrderAmountRequest): OrderAmountResponse {
+        log.debug { "generateOrderAmount: $request" }
+
+        val sale = transaction { MarketplaceSaleEntity[request.saleId] }
+        require(request.bundleQuantity <= sale.availableBundleQuantity) { "Not enough bundles available for sale" }
+
+        val minIncentiveAmount = configRepository.getLong(CONFIG_KEY_MARKETPLACE_MIN_INCENTIVE_AMOUNT)
+        val incentiveAmount =
+            request.incentiveAmount?.also {
+                require(it < minIncentiveAmount) { "Not enough incentive" }
+            } ?: minIncentiveAmount
+
+        val nativeAssets =
+            listOf(
+                nativeAsset {
+                    policy = sale.costPolicyId
+                    name = sale.costAssetName
+                    amount = (sale.costAmount.toBigInteger() * request.bundleQuantity.toBigInteger()).toString()
+                },
+                nativeAsset {
+                    policy = incentivePolicyId
+                    name = incentiveAssetName
+                    amount = (incentiveAmount.toBigInteger() * BigInteger.TWO).toString()
+                }
+            )
+
+        val lovelace = configRepository.getLong(CONFIG_KEY_MARKETPLACE_ORDER_LOVELACE)
+        val amount =
+            CborArray.create(
+                listOf(
+                    CborInteger.create(lovelace.toBigInteger()),
+                    nativeAssets.toNativeAssetCborMap()
+                )
+            )
+
+        val order =
+            transaction {
+                MarketplacePendingOrderEntity.deleteAllExpired(pendingOrderTimeToLive)
+                MarketplacePendingOrderEntity.new {
+                    saleId = sale.id
+                    bundleQuantity = request.bundleQuantity
+                    this.incentiveAmount = incentiveAmount
+                }
+            }
+
+        return OrderAmountResponse(
+            orderId = order.id.value,
+            amountCborHex = amount.toCborByteArray().toHexString()
+        )
+    }
+
+    override suspend fun generateOrderTransaction(request: OrderTransactionRequest): OrderTransactionResponse {
+        log.debug { "generateOrderTransaction: $request" }
+
+        if (request.utxos.isEmpty()) {
+            throw HttpPaymentRequiredException("Missing UTXO's")
+        }
+
+        val (order, sale) =
+            transaction {
+                MarketplacePendingOrderEntity.deleteAllExpired(pendingOrderTimeToLive)
+                MarketplacePendingOrderEntity[request.orderId].let { it to MarketplaceSaleEntity[it.saleId] }
+            }
+
+        val lovelace = configRepository.getLong(CONFIG_KEY_MARKETPLACE_ORDER_LOVELACE)
```

Comment:
Quinn came up with a computation for the wort possible lovelace. So, we'll stick with that value in the database for now.

Reply:
I believe our stream tokens are all worst-case scenario in size anyway, so this should be fine. All of our sample sale tokens are smaller, but this is ok.